### ISSUE TITLE
Switching between projects changes settings and forms

### ISSUE THREAD

**Issue by seadowg** [getodk/collect#4417]
Blocked by #4435, #4453, #4467, #4420, #4498, #4418

## Acceptance

- [x] Given I've changed some settings in Collect
And downloaded and filled out some forms
When I navigate to the settings/projects dialog
And I add a project
And I select it
Then the settings I changed are now the default again
And there are no downloaded or filled out forms
When I switch back to the previous project
Then I see the forms and settings as they were originally

- [x] Given I've installed Collect for the first time on a device
When I opt to use the demo project
Then my server is the demo one (sandbox)
And all my settings are the default

- [x] Given I'm using an older version of Collect
When I upgrade to the new version
Then my settings (including server) are the same as they were before
And my forms are still the same as they were before

## Notes

App data should now be stored by project. The directory structure for a one project setup (or a demo/migrated setup) should look like this:

```
org.odk.collect.android
|-  files
    |-  projects 
        |-  <project uuid>
            |-  forms
            |-  instances
            |-  layers
            |-  metadata
```

`recordings` (and any other ancillary data dirs) will still live at the top level (in `org.odk.collect.android/files`).

**Comment by lognaturel** [comment]
As we're nearing this, I'd love to be looped in on any decisions forming around directory structure. Some decision points I see:
- are we keeping root `forms/`/`instances/`, etc dirs for the first project created? Moving what's there in the case of Collect being updated on a device with an existing configuration?
- are the project directories directly in the root? In a `projects/` subdirectory?
- are the `layers/` and `.cache/` folders shared between projects?
- are the project directories named by their uuid? Numerically increasing? Hash of some identifier (probably not because everything other than UUID can be changed)

Considerations:
- harmonizing with eventual changes for entity-based data collection. This requires a bit of a crystal ball but we can likely make good bets on where we're going.
- `adb` for direct file management. Maybe it doesn't matter too much because of autocomplete.

**Comment by seadowg** [comment]
@grzesiek2010 were just chatting about this! My preference is to have everything nested by project so a form's path (within our app's data directory) would be `<project-uuid>/forms/<form-name>.xml`. That way we don't end up with any special cases which I think make the implementation and any docs (for `adb` access) more confusing. A disadvantage would be that existing users who don't care about projects would have to for `adb` access (although there would only be one project directory). I think an existing install should have its data migrated to that structure as well (a new project is created and everything is moved there). I guess if we want `layers` and `.cache` to be non-project specific then it makes sense to add a `projects` dir to nest these under, but I don't have a strong preference on that either way.  

@lognaturel @grzesiek2010 thoughts?

**Comment by grzesiek2010** [comment]
> I guess if we want layers and .cache to be non-project specific

What's the advantage of this approach? From my perspective, it would be easier to duplicate the same structure of dirs for every project inclusion the dirs you mentioned so that we access all files in the same way, using uuid in StoragePathProvider for example without any exceptions.

**Comment by lognaturel** [comment]
> if we want layers and .cache to be non-project specific

That's a good consideration, thanks. Added to the list.

> What's the advantage of this approach?

This gets to some pretty deep questions about the meaning of projects. 😅 I don't care about `.cache`. There'd be some advantage to having a single cache file if two projects have the same form but the benefits are really minimal.

`layers/` is more subtle, though. In some contexts there'd be huge benefit to having them be shared and it would be a real pain if they're not. First, the only way to get offline tiles onto devices currently is by manually pushing them. Figuring out the uuid for the right project would be a real pain. 

Second, I think projects will often represent different facets of the same work. For example, maybe you have two projects to track two different kinds of assets in the same location. Or you might have two projects because sometimes you're in a data collection role and sometimes you're in a supervisor role for the same context. I'm currently thinking it makes most sense to share layers between projects. I wonder whether there's a way to meet that goal without adding complexity by doing something like having [a symlink](https://developer.android.com/reference/android/system/Os#symlink(java.lang.String,%20java.lang.String)) to a shared layers directory in each project's directory?

More thinking needed. It might be a good idea to chat about this.

**Comment by seadowg** [comment]
> I'm currently thinking it makes most sense to share layers between projects. I wonder whether there's a way to meet that goal without adding complexity by doing something like having a symlink to a shared layers directory in each project's directory?

I think we want most of the code to have no idea that there are different `forms` dirs, so it can be up to `StoragePathProvider` to just always return the same `layers`. I guess docs wise it might be confusing though...

**Comment by lognaturel** [comment]
@seadowg and I briefly discussed and he pointed out that since `StoragePathProvider` abstracts away the filesystem, we can do both and have a shared `layers/` folder whose contents are included for all projects AND special `layers/` folders for each project. That gives users maximum flexibility and likely doesn't introduce too much more complexity. He'll follow up with more details and maybe split it into a separate issue.

In the mean time, here's another consideration -- should project switching be allowed while form filling? My instinct is that it's easier to disable all other profiles in the list. Other options include allowing the switch and just relying on the savepoint for users to get back to where they were (kind of scary given https://github.com/getodk/collect/issues/3494) or showing a dialog similar to the one we show when the user taps the device back button (save, ignore, etc).

**Comment by lognaturel** [comment]
Sigh. I wrote a confusing comment and then deleted it, and then realized that might have been more confusing because it probably landed in your inboxes. 

@seadowg will write up some updates regarding some ideas he shared with me about `layers/`.

I was thinking about project switching during form entry but I realized that we're probably just going to keep the existing overflow menu there for now, right?

**Comment by seadowg** [comment]
> I was thinking about project switching during form entry but I realized that we're probably just going to keep the existing overflow menu there for now, right?

That's what I was assuming yeah.

@seadowg will write up some updates regarding some ideas he shared with me about layers/.

Yup. So we're thinking that we go with having a `layers` and `.cache` in each project's directory as this allows us to do things like encrypt these as part of the project in the future. The one caveat to this is that some setups would probably end up with layers that are added to the device, but that are intended to be shared between projects. To prevent users having to copy their layers to every `layers` dir we'll also add a `layers` directory at the top level. Layers in there will be passed to maps along with whatever is in the current project's `layers` dir. I'll write up an issue for the shared `layers` dir feature so this issue doesn't need to worry about it. I'll add notes to the issue with all this but feel free to continue to chime in if you disagree @grzesiek2010!
